File: crates/rx_bevy_operator_multicast/src/connectable.rs
use rx_bevy_observable::{Observable, ObservableOutput, Observer, Subscription, SubscriptionLike};

pub trait Connectable: Observable {
	fn connect(&mut self) -> Subscription;
}

pub trait SubjectLike: Clone + Observable + Observer + SubscriptionLike {}

pub struct ConnectableObservableConfiguration<Connector, ConnectorCreator>
where
	ConnectorCreator: Fn() -> Connector,
	Connector: 'static + SubjectLike,
{
	connector_creator: ConnectorCreator,
}

/// TODO: Should be part of core or its own
pub struct ConnectableObservable<Source, Connector>
where
	Source: Observable,
	Connector: 'static + SubjectLike<In = Source::Out, InError = Source::OutError>,
{
	/// Upon connection, the connector subject will subscribe to this source
	/// observable
	source: Source,
	/// Upon subscription this connector subject is what will be used as the
	/// source
	connector: Connector,
}

impl<Source, Connector> ObservableOutput for ConnectableObservable<Source, Connector>
where
	Source: Observable,
	Connector: 'static + SubjectLike<In = Source::Out, InError = Source::OutError>,
{
	type Out = Connector::Out;
	type OutError = Connector::OutError;
}

impl<Source, Connector> Observable for ConnectableObservable<Source, Connector>
where
	Source: Observable,
	Connector: 'static + SubjectLike<In = Source::Out, InError = Source::OutError>,
{
	fn subscribe<Destination: 'static + Observer<In = Self::Out, InError = Self::OutError>>(
		&mut self,
		destination: Destination,
	) -> Subscription {
		self.connector.subscribe(destination)
	}
}

impl<Source, Connector> Connectable for ConnectableObservable<Source, Connector>
where
	Source: Observable,
	Connector: 'static + SubjectLike<In = Source::Out, InError = Source::OutError>,
{
	fn connect(&mut self) -> Subscription {
		self.source.subscribe(self.connector.clone())
	}
}

File: crates/rx_bevy_observable_connectable/control/connectable.ts
#!/usr/bin/env bun

import { connectable, finalize, Subject } from "rxjs";

const source = new Subject<number>();

const connectableObservable = connectable(
  source.pipe(finalize(() => console.log("source finalize"))),
  {
    connector: () => {
      console.log("create connector");
      return new Subject<number>();
    },
    resetOnDisconnect: true,
  }
);

source.next(1);

const subscription = connectableObservable
  .pipe(finalize(() => console.log("connection finalize 0")))
  .subscribe({
    next: (next) => console.log("connectable observable 0 - next", next),
    complete: () => console.log("connectable observable 0 - complete"),
  });

console.log("connect 0");
const connection = connectableObservable.connect();

source.next(2);

connection.unsubscribe();
const subscription2 = connectableObservable
  .pipe(finalize(() => console.log("connection finalize 1")))
  .subscribe({
    next: (next) => console.log("connectable observable 1 - next", next),
    complete: () => console.log("connectable observable 1 - complete"),
  });

source.next(3);

console.log("connect 1 (noop)");
connectableObservable.connect();

source.next(4);
